File: course_2/module_1/examples/dfs.py
class Graph:
    def __init__(self):
        # Initialize the graph as an adjacency list.
        self.adjacency_list = {}

    def add_edge(self, u, v, directed=False):
        """
        Add an edge between nodes u and v.
        For undirected graphs, the edge is added in both directions.

        Args:
            u: The starting node of the edge.
            v: The ending node of the edge.
            directed: Boolean indicating whether the edge is directed.
        """
        self.adjacency_list.setdefault(u, []).append(v)
        if not directed:
            self.adjacency_list.setdefault(v, []).append(u)

    def dfs(self, start):
        """
        Perform Depth-First Search (DFS) starting from the node `start`.
        Uses an iterative stack-based approach.

        Args:
            start: The starting node for DFS.

        Returns:
            visited_order: A list of nodes in the order they are visited.
        """
        if not self.adjacency_list:
            raise ValueError("Graph is empty.")
        if start not in self.adjacency_list:
            raise ValueError(f"Start node {start} does not exist in the graph.")

        visited = set()  # Set to track visited nodes.
        visited_order = []  # List to store the order of visited nodes.
        stack = [start]  # Use a stack for iterative DFS.

        while stack:
            node = stack.pop()  # Take the last node added to the stack.
            if node not in visited:
                visited.add(node)  # Mark the node as visited.
                visited_order.append(node)  # Add the node to the visit order.
                # Add unvisited neighbors to the stack.
                stack.extend(neighbor for neighbor in self.adjacency_list.get(node, []) if neighbor not in visited)

        return visited_order

    def _dfs_collect(self, node, visited, component):
        """
        Helper function to collect nodes in a connected component using DFS.

        Args:
            node: The current node being explored.
            visited: Set of already visited nodes.
            component: List to store the current connected component.
        """
        visited.add(node)  # Mark the node as visited.
        component.append(node)  # Add the node to the current component.

        for neighbor in self.adjacency_list.get(node, []):  # Explore neighbors.
            if neighbor not in visited:  # If the neighbor is not visited.
                self._dfs_collect(neighbor, visited, component)  # Recursively explore the neighbor.

    def connected_components(self):
        """
        Identify all connected components in the graph.
        Returns a list of connected components, where each component is a list of nodes.

        Returns:
            components: A list of lists, where each inner list represents a connected component.
        """
        if not self.adjacency_list:
            raise ValueError("Graph is empty.")

        visited = set()  # Set to track visited nodes.
        components = []  # List to store all connected components.

        for node in self.adjacency_list:  # Iterate through all nodes in the graph.
            if node not in visited:  # Start a new DFS if the node is unexplored.
                component = []  # List to store the current connected component.
                self._dfs_collect(node, visited, component)
                components.append(component)  # Add the collected component to the list.

        return components


if __name__ == "__main__":
    # Example usage of the Graph class.
    g = Graph()
    g.add_edge(1, 2)
    g.add_edge(2, 3)
    g.add_edge(4, 5)
    g.add_edge(6, 7)
    g.add_edge(7, 8)
    g.add_edge(8, 6)

    print("Depth-First Search (DFS) Order:")
    print(g.dfs(1))  # Perform DFS starting from node 1.

    print("\nConnected Components:")
    print(g.connected_components())  # Find all connected components in the graph.

File: course_2/module_1/examples/dfs.go
package main

import (
	"fmt"
	"sort"
)

// DFSGraph represents a graph using an adjacency list.
type DFSGraph struct {
	adjacencyList map[int][]int
}

// NewDFSGraph creates and initializes a new DFSGraph.
func NewDFSGraph() *DFSGraph {
	return &DFSGraph{
		adjacencyList: make(map[int][]int),
	}
}

// AddEdge adds an edge between two nodes.
// If the graph is undirected, it adds the edge in both directions.
func (g *DFSGraph) AddEdge(u, v int, directed bool) {
	g.adjacencyList[u] = append(g.adjacencyList[u], v)
	if !directed {
		g.adjacencyList[v] = append(g.adjacencyList[v], u)
	}
}

// DFS performs Depth-First Search starting from the given start node.
// It returns the order of visited nodes.
func (g *DFSGraph) DFS(start int) []int {
	visited := make(map[int]bool) // Track visited nodes.
	result := make([]int, 0)      // Store the order of visited nodes.
	stack := []int{start}         // Use a stack for iterative DFS.

	for len(stack) > 0 {
		// Pop the last node from the stack.
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if !visited[node] {
			visited[node] = true          // Mark the node as visited.
			result = append(result, node) // Add it to the visited order.

			// Add neighbors to the stack in sorted order.
			neighbors := g.adjacencyList[node]
			sort.Ints(neighbors)
			for i := len(neighbors) - 1; i >= 0; i-- {
				if !visited[neighbors[i]] {
					stack = append(stack, neighbors[i])
				}
			}
		}
	}

	return result
}

// ConnectedComponents finds all connected components in the graph.
// It returns a slice of slices, where each inner slice represents a connected component.
func (g *DFSGraph) ConnectedComponents() [][]int {
	visited := make(map[int]bool)  // Track visited nodes.
	components := make([][]int, 0) // Store all connected components.

	for node := range g.adjacencyList {
		if !visited[node] {
			component := make([]int, 0) // Store the current connected component.
			stack := []int{node}

			for len(stack) > 0 {
				// Pop the last node from the stack.
				current := stack[len(stack)-1]
				stack = stack[:len(stack)-1]

				if !visited[current] {
					visited[current] = true                // Mark the node as visited.
					component = append(component, current) // Add to the current component.

					// Add neighbors to the stack in sorted order.
					neighbors := g.adjacencyList[current]
					sort.Ints(neighbors)
					for i := len(neighbors) - 1; i >= 0; i-- {
						if !visited[neighbors[i]] {
							stack = append(stack, neighbors[i])
						}
					}
				}
			}

			components = append(components, component)
		}
	}

	return components
}

func main() {
	graph := NewDFSGraph()
	graph.AddEdge(1, 2, false)
	graph.AddEdge(2, 3, false)
	graph.AddEdge(4, 5, false)
	graph.AddEdge(6, 7, false)
	graph.AddEdge(7, 8, false)
	graph.AddEdge(8, 6, false)

	// Perform DFS starting from node 1.
	fmt.Println("Depth-First Search (DFS) Order:")
	dfsOrder := graph.DFS(1)
	fmt.Println(dfsOrder)

	// Find all connected components in the graph.
	fmt.Println("\nConnected Components:")
	components := graph.ConnectedComponents()
	for i, component := range components {
		sort.Ints(component) // Sort each component for consistent output.
		fmt.Printf("Component %d: %v\n", i+1, component)
	}
}
